### ISSUE TITLE
Accessibility items which didn't make Surge XT 1.0

### ISSUE THREAD

**Issue by baconpaul** [surge-synthesizer/surge#5714]
## 1.1 Beta Problems
- [ ] Typeahead
     - [ ] The typeahead list works great on Mac, if I downarrow into the results VO reads them beautifully. 
     - [ ] On Windows, when I downarrow into the list, the screen readers think that the list itself is focused and not one of the items and so nothing apart from "list" is spoken. 
     - [ ] Also, if I move through the objects with screen reader keys, On Mac the rows just show up as the patch names which is great, while on Windows the cells are first labeled as "row X" and only if you go inside them you can see the label. 
     - [ ] Don't know if this is why Windows doesn't pick up the focus but it could be a clue maybe?

## Outside of 1.1 most likely

- [ ] Lastly, I'm not sure if this isn't a Juce problem but would it be possible to make it so that hovering the mouse over a submenu wouldn't expand it or at the very least make the focus not go to it if you open a menu from the keyboard? VoiceOver on Mac has an option where the VO cursor and mouse pointer are synchronized, and this is useful on many web sites that hide information unless you hover over something. What happens if this is on and you go into the main menu in Surge is the Zoom menu gets focus, VoiceOver and the mouse jumps to it which causes it to open and focus the "100%" option. If you try to left arrow out of the menu the focus will go back for an instant but then bounce back in trapping you in unless you temporarily turn off cursor tracking.
- [ ] MSEG control area, but we are tracking that separately in #5934 and it may not make it to 1.1.

**Comment by mkruselj** [comment]
Shift+up/down on sequencer steps should perhaps match what Shift+drag does with mouse (which is quantize to semitones)? Or we move that shortcut to Alt (but this would be inconsistent then).

**Comment by pitermach** [comment]
This will probably be long but I figured I'll write my thoughts down before I forget something. ☺️ Here's some more thoughts after using Surge more as well as some Twitter conversations (reception of XT 1.0 has been very positive so far!)

Regarding Step Sequencer, I think I like the triggers being next to the steps in a way because it's easier to set them up in the moment as you program a sequence. However, if you go with the screen reader cursor they show up before the steps. I think we should get feedback from more people which workflow they prefer and then make the 2 match.

For patch browser, type ahead and also the category auto complete when saving the patch, the main issue is the list items don't speak their text. They can be navigated to and arrowed through, but all screen readers say is "row 1, row 2..." and so on. 

Specifically for the patch browser, while you can arrow through the category tree and patch list, you can't load a patch either by pressing enter or doing a press action with a screen reader, both should probably work.

Regarding CTRL+Tabbing through groups, in the previous issue we also discussed having dedicated hotkeys to jump to specific groups (I think we settled on CTRL+Shift+Letters as good candidates), I think both features are worth doing.

I was also thinking it might be useful to speak some announcements for some actions as a result from hotkeys being pressed. The main one being telling you when a patch was added or removed from favorites when you hit Alt+F because I got confused a few times when going through all the patches and batch marking as favorites. Speaking the patch name when you switch with Right/Left might also be useful.

Pressing space/enter should cycle the state of toggles and press buttons (like the oscillator type), this is for consistency with how these controls work in the OS. Some buttons actually already respond to enter (like the Close button in the modulation list or the OK and Cancel buttons in alerts.

A few points in menus:
- The extra items added that let you edit modulations are skipped over when you arrow through the menu. You can activate them with a screen reader if you know they're there but they're easy to miss if you don't.
- Enter should go into submenus in addition to right arrow. Right now, you can only enter a submenu by pressing right and activate items with Enter.
- In a similar vane to the checked status not being natively reported, there is no "sub menu" indication on Mac, though interestingly enough there is on Windows. This is probably a Juce thing because IIRC I've seen it in Pianoteq also.
- Lastly, I'm not sure if this isn't a Juce problem but would it be possible to make it so that hovering the mouse over a submenu wouldn't expand it or at the very least make the focus not go to it if you open a menu from the keyboard? VoiceOver on Mac has an option where the VO cursor and mouse pointer are synchronized, and this is useful on many web sites that hide information unless you hover over something. What happens if this is on and you go into the main menu in Surge is the Zoom menu gets focus, VoiceOver and the mouse jumps to it which causes it to open and focus the "100%" option. If you try to left arrow out of the menu the focus will go back for an instant but then bounce back in trapping you in unless you temporarily turn off cursor tracking.

**Comment by baconpaul** [comment]
All super useful.

Some things (like intersperse triggers) is easy. Other things may be harder.

I'm glad it is usable in 1.0 and am super excited to hone it for 1.1 :)

**Comment by baconpaul** [comment]
Oh also where on twitter does the convo happen out of curiosity?

**Comment by pitermach** [comment]
https://twitter.com/OnjMusic/status/1484097351157436416

**Comment by mkruselj** [comment]
- [ ] Pressing End on mod amount slider in Mod List goes to 3x the range. For example, if minimum mod amount is -100%, it will go to -300%. If it's -14 semitones for pitch modulation, it goes to -42 st.
- [ ] Using up/down keys on mod amount slider doesn't produce useful steps (i.e. it jumps in 1.14 st for pitch). I reckon these steps should be the same as if we're doing a Ctrl/Cmd+drag - quantized to proper steps.

**Comment by baconpaul** [comment]
Accessibility bug with screen readers and type-in fields, editing is difficult because the announced character is always off by one, so you end up deleting the wrong character or inserting something in the wrong place.

From discord user bcross

**Comment by pitermach** [comment]
Just did some more experimenting on this:

- It seems to be a Juce issue because it also happens in Pianoteq
- It's Windows specific - Mac works totally fine
- You can reproduce it with any screen reader (tried with NVDA and Narrator)

This effects both deleting and cursor movement. If you go to the beginning of a line, then press the right arrow you should hear the screen reader speak the second character, but it repeats the first instead.

I also found a second issue on Windows though this one only seems to effect NVDA, if you make a selection by holding down shift and moving through the text, NVDA always repeats the entire selection, rather than saying just the part that was selected/unselected.

**Comment by baconpaul** [comment]
OK I'll report this over on the juce forums and see if there's a fix we can get put in place from the team (or if the fix is there; i'm a few points behind still and plan to pull to 6.1.5 in this cycle).

**Comment by baconpaul** [comment]
https://forum.juce.com/t/accessibility-bug-with-windows-and-text-entry-widgets/50057

**Comment by pitermach** [comment]
Just noticed another thing, there's currently no way to accessibly edit the "adaptive" shape of the alias oscillator. Manual mentions there should be an Edit button when it's selected, but I don't see one with VO.

**Comment by baconpaul** [comment]
oh yeah that's not exposed! Good catch.

**Comment by ironcross32** [comment]
- [ ] Using arrows to traverse oscillator types are not spoken when the type changes.
- [ ] When changing the engine under the Twist oscillator type, if the arrows are used to do this rather than using the context menu, the screen reader does not see the controls as having been updated.

- [ ] Filter subtypes are not announced when using the arrows to traverse them.

**Comment by baconpaul** [comment]
From discord 

> Re the oscillator type not being spoken when you arrow to change it, perhaps if you change its reported control type from a button to a combo box the screen readers would pick up on the value changing. You generally can't arrow up and down on buttons to modify something so maybe they're not even monitoring them for event changes.

**Comment by baconpaul** [comment]
If I apply this diff
```cpp
 case tag_settingsmenu:
    {
        auto h = frame->getAccessibilityHandler();
        if (h)
        {
            std::cout << "GOT HAND" << std::endl;
            h->postAnnouncement("Ouch you poked me",
                                juce::AccessibilityHandler::AnnouncementPriority::high);
        }
        else
        {
            juce::Point<int> menuPoint;
            useDevMenu = false;
            showSettingsMenu(menuPoint, control);
        }
    }
```
then clicking the menu button with the mac screen reader no longer pops up the main menu but instead says "Ouch you poked me"
which is of course not useful
but implies that there's a useful thing to be done
And now I can solve a lot of problems.
I make a std::string member of SGE called "announceNextIdle"
and voila

**Comment by baconpaul** [comment]
@pitermach  says:

If it works for this it might be a good way to make the type ahead speak while you're typing as well. So like you stop typing for a moment and it could say something like "3 results, init saw" so that'd tell you what patch would load to press enter but also that you have other results to look at
That's more or less how auto complete type things are made accessible usually from a design point of view at least

**Comment by baconpaul** [comment]
Just remembering this diff
```
 void MenuCenteredBoldLabel::addToMenu(juce::PopupMenu &m, const std::string label)
 {
-    m.addCustomItem(-1, std::make_unique<MenuCenteredBoldLabel>(label));
+    m.addCustomItem(-1, std::make_unique<MenuCenteredBoldLabel>(label), nullptr, label);
 }
```

**Comment by pitermach** [comment]
Just realized that the on-screen Macro controls aren't exposed as sliders. If you focus one of them in the modulators section and press up/down they adjust, but you have no way to see their value. While up/down work, home/end/delete don't.

**Comment by pitermach** [comment]
I just saw the keyboard shortcut editor mentioned in the nightly changes and took a look at it. Here's a couple of things I noticed:

- The keybinds table according to both Mac and Windows has just 1 column, the name. If I navigate through the rows with VO or NVDA navigation I get the functions spoken, but I don't hear what the key is set to and can't reach the "toggle" checkboxes.
- If I tab through the list, then on Windows screen readers announce the toggles and I can do the screen reader press to change them. On Mac, VOiceOver still just reads the name only and I don't think doing the press action is clicking the right thing
- The 2 other buttons in the window apart from "Close" don't have a label.


Either way this is a really handy feature and the fact this list speaks more makes me happy in terms of things like the patch browser down the road. ☺️

**Comment by baconpaul** [comment]
Yes sorry I mentioned on discord thst I merged this half done before a mini vacation. While I could tab through it I haven’t done accessibility revie (or keybind reset) yet both of which are needed before #5912 is closed

**Comment by pitermach** [comment]
Ah right, I've been a bit busy and didn't look at the Discord in a while and missed that part, sorry. Hope you enjoyed your vacation

**Comment by baconpaul** [comment]
No worries! Just making sure you knew I agreed it wasn’t done :)

**Comment by baconpaul** [comment]
With the windows workaround for speaking and the full resolution moved to #6408 I'm going to close this.
